File: security-jwt/src/test/groovy/io/micronaut/docs/rejection/MyRejectionHandler.java
package io.micronaut.docs.rejection;

//tag::clazz[]
import io.micronaut.context.annotation.Replaces;
import io.micronaut.context.annotation.Requires;
import io.micronaut.core.annotation.Nullable;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.MutableHttpResponse;
import io.micronaut.http.server.exceptions.response.ErrorResponseProcessor;
import io.micronaut.security.authentication.AuthorizationException;
import io.micronaut.security.authentication.DefaultAuthorizationExceptionHandler;

import io.micronaut.security.config.RedirectConfiguration;
import io.micronaut.security.errors.PriorToLoginPersistence;
import jakarta.inject.Singleton;

//end::clazz[]

@Requires(property = "spec.name", value = "rejection-handler")
//tag::clazz[]
@Singleton
@Replaces(DefaultAuthorizationExceptionHandler.class)
public class MyRejectionHandler extends DefaultAuthorizationExceptionHandler {

    public MyRejectionHandler(
            RedirectConfiguration redirectConfiguration,
            @Nullable PriorToLoginPersistence priorToLoginPersistence,
            @Nullable ErrorResponseProcessor<?> responseProcessor
    ) {
        super(redirectConfiguration, priorToLoginPersistence, responseProcessor);
    }

    @Override
    public MutableHttpResponse<?> handle(HttpRequest request, AuthorizationException exception) {
        //Let the DefaultAuthorizationExceptionHandler create the initial response
        //then add a header
        return super.handle(request, exception).header("X-Reason", "Example Header");
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/security/docs/rejection/MyRejectionHandler.py
# tag::clazz[]
from jakarta.inject import Singleton
from micronaut.context.annotation import Replaces, Requires
from micronaut.http import HttpRequest, MutableHttpResponse
from micronaut.http.server.exceptions import ExceptionHandler
from micronaut.http.server.exceptions.response import ErrorResponseProcessor
from micronaut.security.authentication import AuthorizationException, DefaultAuthorizationExceptionHandler, WwwAuthenticateChallengeProvider
from micronaut.security.config import RedirectConfiguration, RedirectService
from micronaut.security.errors import PriorToLoginPersistence

# end::clazz[]
@Requires(property="spec.name", value="RejectionHandlerOverrideTest")
# tag::clazz[]
@Singleton
@Replaces(DefaultAuthorizationExceptionHandler)
class MyRejectionHandler(ExceptionHandler[AuthorizationException, MutableHttpResponse]):

    def __init__(self,
                 errorResponseProcessor: ErrorResponseProcessor,
                 redirectConfiguration: RedirectConfiguration,
                 redirectService: RedirectService,
                 wwwAuthenticateChallengeProviders: list[WwwAuthenticateChallengeProvider],
                 priorToLoginPersistence: PriorToLoginPersistence | None):
        # A Python class cannot extend a Java class, so delegate to the DefaultAuthorizationExceptionHandler
        self.delegate = DefaultAuthorizationExceptionHandler(errorResponseProcessor, redirectConfiguration, redirectService, wwwAuthenticateChallengeProviders, priorToLoginPersistence)

    def handle(self, request: HttpRequest, exception: AuthorizationException) -> MutableHttpResponse:
        # Let the DefaultAuthorizationExceptionHandler create the initial response
        # then add a header
        return self.delegate.handle(request, exception).header("X-Reason", "Example Header")
# end::clazz[]
